### PR TITLE
[WIP] Synchronous temporal filter workflow

### PR DESCRIPTION
**Pull request by rciric**
This PR addresses issue #311.

* A generalised temporal filtering workflow applies identical operations to any number of 4D (NIfTI) and 2D (TSV) timeseries provided as input, enabling [synchronous filtering of confounds and data](https://www.ncbi.nlm.nih.gov/pubmed/23747457).
* Available filters are derived from scipy (Butterworth, Chebyshev, and elliptic filters), AFNI (FFT-based filter), and FSL (Gaussian kernel convolutional filter).
* To fully enable the latter filters, an additional workflow is included for applying any 4D image processing function to 2D data. The AFNI arm of that workflow leverages existing functionality within AFNI (see [1D files as AFNI datasets here](https://afni.nimh.nih.gov/pub/dist/doc/program_help/common_options.html)); however, its implementation is rather hackish (as Nipype doesn't allow 1d as a valid AFNI `outputtype`) and should potentially be brought up as a Nipype issue.
* (Note: if you're using a fit-based filter -- for instance with the cosine basis functions provided with fmriprep -- you probably should be doing single-pass filtering and regression rather than using a workflow like this.)
* The filtering workflow optionally incorporates a demean/detrend step and an interpolation step to reduce the impact of high-motion frames. Despiking might also be worth considering.

Marking as WIP, as more robust testing still needs to be performed and appropriate unit tests probably added.

### CODE REVIEW

**Comment by oesteban** [comment]
Hey @kfinc, would you be interested in moving this filtering into fMRIDenoise? WDYT @rciric ?

**Comment by kfinc** [comment]
@oesteban @rciric could you let us know specifically what's the difference between `tfilter` and filter applied in `nilearn.signal.clean` (https://nilearn.github.io/modules/generated/nilearn.signal.clean.html)?

**Comment by rciric** [comment]
When time series data with censored frames are temporally filtered, signal information from the censored frames can blur into adjacent frames (see [Carp 2013](https://www.ncbi.nlm.nih.gov/pubmed/22227884)). (This problem likely affects autoregressive filters like Butterworth, but intuitively, I’m not sure that it would occur if a DCT basis is used with spike regressors.)

The `tfilter` routine differs from `nilearn.signal.clean` in that it addresses this particular problem (at the cost of additional computational steps). It uses a transform based on the Lomb-Scargle periodogram to implement an interpolation over censored frames prior to application of the filter. This enables implementation of the “iterative” processing (or “reprocessing”) scheme recommended by [Power and colleagues](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3849338). `tfilter` furthermore supports a broader class of filters (Butterworth — currently used by `nilearn.signal.clean`, Gaussian kernel, Chebyshev I/II, elliptic, and FFT-based). (DCT filter requires no specific support, as the basis can be included as part of the confound model during the regression step; I believe that this is the approach implemented in AFNI’s `3dTproject`. Also, `tfilter` does currently depend on FSL and AFNI to implement some filter classes.) It looks like nilearn also standardizes confound ranges to improve numerical stability.

See Figure S9 in the [Power supplement](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3849338) for an example of the utility of the interpolation step.

**Comment by oesteban** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 43
           

Complexity increasing per file
==============================
- niworkflows/func/timeseries1D.py  2
- niworkflows/interfaces/timeseries1D.py  3
- niworkflows/func/filter.py  7
- niworkflows/interfaces/filter.py  7
         

Clones added
============
- niworkflows/interfaces/filter.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913&bid=16766271)

File: niworkflows/func/filter.py [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+
+    workflow.connect([
+        (src_4d[0], dst_4d[0], [(src_4d[1], dst_4d[1])]),
+        (src_2d[0], dst_2d[0], [(src_2d[1], dst_2d[1])]),
+    ])
+    return workflow
+
+
+def _validate_passband(filter_type, passband):
+    """Verify that the filter passband is reasonably formulated.
```
Comment (oesteban) [inline niworkflows/func/filter.py]: Codacy Issue found: [Missing blank line after last section ('Parameters')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 2]
```diff
@@ -0,0 +1,1165 @@
...
+        passband_norm = passband_norm[0]
+    elif filter_pass == 'lowpass':
+        passband_norm = passband_norm[1]
+    elif passband_norm[0] > passband_norm[1]:
+        filter_pass = 'bandstop'
+    return passband_norm, filter_pass
+
+
+def _unfold_image(img, mask=None):
+    """Unfold a four-dimensional time series into two dimensions.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing dashed underline after section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+        5000, usedefault=True, desc='Number of voxels to transform at a time')
+    output_file = File(desc='Output path')
+
+
+class Interpolate4DOutputSpec(TraitedSpec):
+    output_file = File(exists=True, desc='Interpolated 4D NIfTI file')
+
+
+class Interpolate4D(SimpleInterface):
+    """Interpolate a 4D NIfTI time series.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Multi-line docstring summary should start at the second line](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 4]
```diff
@@ -0,0 +1,1165 @@
...
+    return timeseries_4d_out
+
+
+def interpolate_lombscargle_2d(timeseries_2d,
+                               timeseries_2d_out,
+                               t_rep,
+                               temporal_mask=None,
+                               oversampling_frequency=8,
+                               maximum_frequency=1):
+    """Interpolation for 2D TSV time series data.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 5]
```diff
@@ -0,0 +1,1165 @@
...
+                '==================================================\n'
+                'Sampling rate:     {0}\n'
+                'Nyquist frequency: {1}\n'
+                'High-pass cutoff:  {2[0]}\n'
+                'Low-pass cutoff:   {2[1]}\n'.format(
+                    sampling_rate, nyquist, passband))
+        passband_norm[0] = 0
+        filter_pass = 'lowpass'
+
+    if filter_pass == 'highpass':
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Possible hardcoded password: 'highpass'](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 6]
```diff
@@ -0,0 +1,1165 @@
...
+        desc='Order of polynomial detrend (0 for demean only)')
+    output_file = File(desc='Output path')
+
+
+class DemeanDetrend2DOutputSpec(TraitedSpec):
+    output_file = File(exists=True, desc='Demeaned and detrended TSV file')
+
+
+class DemeanDetrend2D(SimpleInterface):
+    """Demean/detrend a 2D TSV time series."""
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [1 blank line required after class docstring (found 0)](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 7]
```diff
@@ -0,0 +1,1165 @@
...
+    idx_nan = traits.List(desc='List of each column\'s NaN indices')
+    output_file = File(desc='Output path')
+
+
+class RestoreNaNOutputSpec(TraitedSpec):
+    output_file = File(exists=True, desc='TSV file with NaNs restored')
+
+
+class RestoreNaN(SimpleInterface):
+    """Restore NaN values that were interpolated over."""
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [1 blank line required after class docstring (found 0)](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 8]
```diff
@@ -0,0 +1,1165 @@
...
+
+
+def interpolate_lombscargle(data,
+                            sine_term,
+                            cosine_term,
+                            angular_frequencies,
+                            all_samples,
+                            n_samples_seen,
+                            n_samples):
+    """Temporally interpolate over unseen (masked) values in a dataset using
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/func/filter.py [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+            'High-pass cutoff must be less than low-pass cutoff\n'
+            'for the selected filter type.\n'
+            '==================================================\n'
+            'Filter class:     {0}\n'
+            'High-pass cutoff: {1[0]}\n'
+            'Low-pass cutoff:  {1[1]}\n'.format(filter_type, passband))
+
+
+def _get_fsl_passband(passband, sampling_rate):
+    """
```
Comment (oesteban) [inline niworkflows/func/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 10]
```diff
@@ -0,0 +1,1165 @@
...
+                      timeseries_4d_out,
+                      brain_mask=None,
+                      t_rep=None,
+                      filter_type='butterworth',
+                      filter_order=1,
+                      passband=(0.01, 0.08),
+                      ripple_pass=5,
+                      ripple_stop=20):
+
+    """Temporally filter a 4D NIfTI dataset.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/func/timeseries1D.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+
+def init_apply_to_2d_wf(function_node,
+                        t_rep,
+                        name='apply_to_2d_wf',
+                        mode=None,
+                        input_field='in_file',
+                        output_field='out_file'):
+    """
```
Comment (oesteban) [inline niworkflows/func/timeseries1D.py]: Codacy Issue found: [First word of the docstring should not be `This`](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    tsv_data.shape = [tsv_data.shape[0], 1, 1, tsv_data.shape[1]]
+    header = list(pd.read_csv(tsv, sep='\t', nrows=0, header=0).columns)
+    img_data = nb.Nifti1Image(dataobj=tsv_data, affine=np.eye(4))
+    img_data.header['pixdim'][4] = t_rep
+    nb.save(img_data, img)
+    return img, header
+
+
+def img2tsv(img, tsv, mask=None, header=None):
+    """Dump data from an image file into tabular form.
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+    tsv_data = pd.read_csv(tsv, sep='\t', header=None, skiprows=[0])
+    tsv_header = list(pd.read_csv(tsv, sep='\t', nrows=0, header=0).columns)
+    if transpose:
+        tsv_data = tsv_data.T
+    tsv_data.to_csv(afni1d, sep='\t', index=False, na_rep='n/a', header=False)
+    return afni1d, tsv_header
+
+
+def tsv_from_1D(afni1d, tsv, obs='columns', header=None):
+    """Convert an AFNI 1D file into a tsv file. Note that header information
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 14]
```diff
@@ -0,0 +1,1165 @@
...
+                'Nyquist frequency: {1}\n'
+                'High-pass cutoff:  {2[0]}\n'
+                'Low-pass cutoff:   {2[1]}\n'.format(
+                    sampling_rate, nyquist, passband))
+        passband_norm[0] = 0
+        filter_pass = 'lowpass'
+
+    if filter_pass == 'highpass':
+        passband_norm = passband_norm[0]
+    elif filter_pass == 'lowpass':
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Possible hardcoded password: 'lowpass'](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/func/__init__.py [hunk 15]
```diff
@@ -2,3 +2,4 @@
 # emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
 # vi: set ft=python sts=4 ts=4 sw=4 et:
 from __future__ import absolute_import, division, print_function, unicode_literals
+from .filter import init_temporal_filter_wf
```
Comment (oesteban) [inline niworkflows/func/__init__.py]: Codacy Issue found: ['.filter.init_temporal_filter_wf' imported but unused](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 16]
```diff
@@ -0,0 +1,1165 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+"""
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [First line should end with a period, question mark, or exclamation point (not 's')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 17]
```diff
@@ -0,0 +1,1165 @@
...
+
+
+def general_filter(data,
+                   sampling_rate,
+                   filter_type='butterworth',
+                   filter_order=1,
+                   passband=(0.01, 0.08),
+                   ripple_pass=5,
+                   ripple_stop=20):
+    """Temporal filtering for any data array.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 18]
```diff
@@ -0,0 +1,1165 @@
...
+
+def interpolate_lombscargle_4d(timeseries_4d,
+                               timeseries_4d_out,
+                               brain_mask=None,
+                               temporal_mask=None,
+                               t_rep=None,
+                               oversampling_frequency=8,
+                               maximum_frequency=1,
+                               voxel_bin_size=5000):
+    """Interpolation for 4D NIfTI time series data.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+        self._results['out_file'] = tsv_from_1D(afni1d=self.inputs.in_file,
+                                                tsv=out_file,
+                                                obs=self.inputs.in_format,
+                                                header=self.inputs.header)
+
+        return runtime
+
+
+def tsv2img(tsv, img, t_rep):
+    """Create a dummy NIfTI image out of a tsv file, thus enabling it to be
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 20]
```diff
@@ -0,0 +1,262 @@
...
+    t_rep = traits.Float(mandatory=True, desc='Repetition time')
+
+
+class TSVToImageOutputSpec(TraitedSpec):
+    out_file = File(exists=True, desc='Image derived from input TSV')
+    header = traits.CList(traits.Str, desc='Header stripped from TSV')
+
+
+class TSVToImage(SimpleInterface):
+    """Convert a TSV into an image"""
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [First line should end with a period, question mark, or exclamation point (not 'e')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 21]
```diff
@@ -0,0 +1,1165 @@
...
+    output_mean = File(desc='Output path for mean image')
+
+
+class DemeanDetrend4DOutputSpec(TraitedSpec):
+    output_file = File(exists=True, desc='Demeaned and detrended NIfTI file')
+    output_mean = File(exists=True, desc='Mean value computed voxelwise')
+
+
+class DemeanDetrend4D(SimpleInterface):
+    """Demean/detrend a 4D NIfTI time series."""
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [1 blank line required after class docstring (found 0)](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 22]
```diff
@@ -0,0 +1,1165 @@
...
+
+
+def demean_detrend_4d(timeseries_4d,
+                      timeseries_4d_out,
+                      detrend_order,
+                      brain_mask=None,
+                      temporal_mask=None,
+                      save_mean=False,
+                      mean_out=None):
+    """Demean and detrend a 4D NIfTI data set.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/filter.py [hunk 23]
```diff
@@ -0,0 +1,1165 @@
...
+    """
+    if mask is not None:
+        mask = nb.load(mask)
+        return img.get_fdata()[mask.get_data().astype('bool')]
+    else:
+        return img.get_fdata().reshape([-1, img.shape[3]])
+
+
+def _fold_image(data, template, mask=None):
+    """Fold a 2D numpy array into a 4D NIfTI time series.
```
Comment (oesteban) [inline niworkflows/interfaces/filter.py]: Codacy Issue found: [Missing blank line after last section ('Returns')](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 24]
```diff
@@ -0,0 +1,262 @@
...
+    header = traits.Either(None, traits.List, default=None, usedefault=True,
+                           desc='header to prepend to the output TSV')
+
+
+class TSVFrom1DOutputSpec(TraitedSpec):
+    out_file = File(exists=True, desc='TSV derived from input 1D file')
+
+
+class TSVFrom1D(SimpleInterface):
+    """Convert an AFNI 1D file to (potentially) BIDS-compatible TSV format."""
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [1 blank line required after class docstring (found 0)](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

File: niworkflows/interfaces/timeseries1D.py [hunk 25]
```diff
@@ -0,0 +1,262 @@
...
+        desc='NIfTI mask for TSV extraction')
+    header = traits.CList(traits.Str, desc='Header for TSV columns')
+
+
+class ImageToTSVOutputSpec(TraitedSpec):
+    out_file = File(exists=True, desc='TSV derived from input image')
+
+
+class ImageToTSV(SimpleInterface):
+    """Convert an image into a TSV."""
```
Comment (oesteban) [inline niworkflows/interfaces/timeseries1D.py]: Codacy Issue found: [1 blank line required after class docstring (found 0)](https://app.codacy.com/gh/nipreps/niworkflows/pullRequest?prid=5096913)

**Comment by effigies** [comment]
This seems dead. Please rebase and reopen if this is still a goal.

**Pull request closed without merging**